utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/encode/BestEffortJsonSchemaEncoder.scala

```diff
@@ -0,0 +1,92 @@
+  }
+
+  private def encodeObject(fields: Map[String, _], parentSchema: ObjectSchema): EncodeOutput = {
+    fields
+      .map(field => (field, parentSchema.getPropertySchemas.get(field._1)))
+      .collect {
+        case (field, propertySchema) if (propertySchema != null) && notNullOrRequired(field._1, field._2, parentSchema) => encode(field._2, propertySchema).map(field._1 -> _)
+        case (filed, null) if !parentSchema.permitsAdditionalProperties() => error(s"Not expected field with name: ${filed._1} for schema: $parentSchema and policy $validationMode does not allow redundant")
+      }
+      .toList.sequence.map { values => Json.fromFields(values) }
+  }
+
+  private def notNullOrRequired(fieldName: String, value: Any, parentSchema: ObjectSchema): Boolean = {
+    value != null || parentSchema.getRequiredProperties.contains(fieldName)
+  }
+
+  private def encodeCollection(collection: Traversable[_], schema: ArraySchema): EncodeOutput = {
+    collection.map(el => encode(el, schema.getAllItemSchema)).toList.sequence.map(l => Json.fromValues(l))
+  }
+
+  def encodeBasedOnSchema(input: EncodeInput): EncodeOutput = {
+    val (value, schema, fieldName) = input
+    (schema, value) match {
+      case (schema: ObjectSchema, map: scala.collection.Map[String@unchecked, _]) => encodeObject(map.toMap, schema)
+      case (schema: ObjectSchema, map: java.util.Map[String@unchecked, _]) => encodeObject(map.toMap, schema)
+      case (schema: ArraySchema, value: Traversable[_]) => encodeCollection(value, schema)
+      case (schema: ArraySchema, value: java.util.Collection[_]) => encodeCollection(value.toArray, schema)
+      case (schema: StringSchema, value: Any) => encodeStringSchema(schema, value, fieldName)
+      case (_: NumberSchema, value: Long) => Valid(jsonEncoder.encode(value))
+      case (_: NumberSchema, value: Double) => Valid(jsonEncoder.encode(value))
+      case (_: NumberSchema, value: Float) => Valid(jsonEncoder.encode(value))
+      case (_: NumberSchema, value: Int) => Valid(jsonEncoder.encode(value))
+      case (_: NumberSchema, value: java.math.BigDecimal) => Valid(jsonEncoder.encode(value))
+      case (_: NumberSchema, value: java.math.BigInteger) => Valid(jsonEncoder.encode(value))
+      case (_: NumberSchema, value: Number) => Valid(jsonEncoder.encode(value.doubleValue()))
+      case (_: NullSchema, null) => Valid(Json.Null)
+      case (_: NullSchema, None) => Valid(Json.Null)
+      case (_: BooleanSchema, value: Boolean) => Valid(Json.fromBoolean(value))
+      case (_: EnumSchema, value: Enum[_]) => Valid(Json.fromString(value.toString))
+      case (null, value: Any) if validationMode == ValidationMode.lax => Valid(jsonEncoder.encode(value))
+      case (_, null) if validationMode != ValidationMode.lax => error(s"Not expected null for field: $fieldName with schema: $schema")
+      case (null, _) if validationMode != ValidationMode.lax => error(s"Not expected null for field: $fieldName with schema: $schema")
```

Comment:
Is it a but? Here you will have always message: Not expected null for field: $fieldName with schema: null. What mean this message? What mean schema: null?

Reply:
yes, this case is redundant